src/cuda/cuda_topk_full_sort.cuh

```diff
@@ -0,0 +1,104 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+#pragma once
+
+#include <cub/device/device_segmented_radix_sort.cuh>
+#include "cuda_topk.h"
+
+namespace Generators {
+namespace cuda {
+
+// Kernel to initialize an index tensor where indices[i] = i % vocab_size.
+__global__ void PopulateIndices(int* indices, int vocab_size, int batch_size);
+
+void LaunchPopulateIndices(int* indices, int vocab_size, int batch_size, cudaStream_t stream);
+
+// Kernel to create the segment offsets required by CUB's segmented sort.
+__global__ void PopulateOffsets(int* offsets, int vocab_size, int batch_size);
+
+void LaunchPopulateOffsets(int* offsets, int vocab_size, int batch_size, cudaStream_t stream);
+
+// Wrapper for CUB's segmented radix sort for (float, int) key-value pairs.
+void LaunchSortPairs(void* d_temp_storage, size_t temp_storage_bytes, const float* d_keys_in, float* d_keys_out,
+                     const int* d_values_in, int* d_values_out, int num_items, int num_segments, int* d_offsets,
+                     cudaStream_t stream);
+
+// Helper to determine the size of temporary storage needed by CUB for the sort.
+inline size_t GetFullSortCubTempStorageBytes(int num_items, int num_segments, cudaStream_t stream);
+
+// Performs a full sort on the entire vocabulary for each batch item, writing to the specified buffers.
+// This is the non-compacting version used internally by the high-performance GetTopK path.
+void LaunchSort(TopkData* data, cudaStream_t stream, const float* scores_in, float* scores_out, int* indices_out,
+                int vocab_size, int batch_size);
+
+// --- Implementations ---
+
+__global__ void PopulateIndices(int* indices, int vocab_size, int batch_size) {
+  int global_index = threadIdx.x + blockIdx.x * blockDim.x;
+  if (global_index < vocab_size * batch_size) {
+    indices[global_index] = global_index % vocab_size;
+  }
+}
+
+inline void LaunchPopulateIndices(int* indices, int vocab_size, int batch_size, cudaStream_t stream) {
+  dim3 grid(CeilDiv(batch_size * vocab_size, 256), 1, 1);
+  dim3 block(256, 1, 1);
+  PopulateIndices<<<grid, block, 0, stream>>>(indices, vocab_size, batch_size);
+  CUDA_CHECK(cudaGetLastError());
+}
+
+__global__ void PopulateOffsets(int* offsets, int vocab_size, int batch_size) {
+  int index = threadIdx.x + blockIdx.x * blockDim.x;
+  if (index < batch_size + 1) {
+    offsets[index] = index * vocab_size;
+  }
+}
+
+inline void LaunchPopulateOffsets(int* offsets, int vocab_size, int batch_size, cudaStream_t stream) {
+  // We need batch_size + 1 offsets for CUB's segmented sort.
+  dim3 grid(CeilDiv(batch_size + 1, 256), 1, 1);
+  dim3 block(256, 1, 1);
+  PopulateOffsets<<<grid, block, 0, stream>>>(offsets, vocab_size, batch_size);
+  CUDA_CHECK(cudaGetLastError());
```

Comment:
Is this needed ? Can this be enabled only on debug builds ?

Reply:
It's needed since the segment radix need to know the offset of each batch.

Reply:
You mean `cudaGetLastError()` is needed ?

Reply:
I updated the check according to your suggestion.